### PR TITLE
Bugfix/fake fee charge

### PR DESCRIPTION
**Pull request by Yogev911**

### CODE REVIEW

File: services/horizon/internal/actions_transaction.go [hunk 1]
```diff
@@ -13,6 +11,7 @@ import (
 	"github.com/kinecosystem/go/services/horizon/internal/txsub"
 	"github.com/kinecosystem/go/support/render/hal"
 	"github.com/kinecosystem/go/support/render/problem"
+	"net/http"
```
Comment (oryband) [inline services/horizon/internal/actions_transaction.go:14]: `net/http` should b ekept where it was. i general keep stdlib imports before 3rd party imports and separated by a line space

File: services/horizon/internal/actions_transaction.go [hunk 2]
```diff
@@ -129,6 +128,10 @@ type TransactionShowAction struct {
 	Resource horizon.Transaction
 }
 
+type WhitelistAccount struct {
+	WhitelistedAccounts map[string]string `json:"data"`
```
Comment (oryband) [inline services/horizon/internal/actions_transaction.go]: if these are the whitelisted accounts i think the json field name should be named `"whilelist_account"` or anything else that makes sense. data doesn't mean anything specific

File: services/horizon/internal/actions_transaction.go [hunk 3]
```diff
@@ -148,11 +151,27 @@ func (action *TransactionShowAction) JSON() error {
...
-		func() { hal.Render(action.W, action.Resource) },
+		func() {
+			validateFee(action)
+			hal.Render(action.W, action.Resource)
+		},
 	)
 	return action.Err
 }
 
+func validateFee(action *TransactionShowAction) {
```
Comment (oryband) [inline services/horizon/internal/actions_transaction.go]: - add function doc briefly explaining what this function does
- "validate fee" name is confusing - what fee is it validating? why does a fee need validation?

File: services/horizon/internal/actions_transaction.go [hunk 4]
```diff
@@ -148,11 +151,27 @@ func (action *TransactionShowAction) JSON() error {
...
+	mAccountAction := AccountShowAction{}
+	mWhiteListAccount := WhitelistAccount{}
+
+	_ = action.CoreQ().AllDataByAddress(&mAccountAction.CoreData, action.App.WhiteListAccount)
+	for _, d := range mAccountAction.CoreData {
+		mWhiteListAccount.WhitelistedAccounts[d.Key] = d.Value
+	}
+	if _, found := mWhiteListAccount.WhitelistedAccounts[action.Resource.Account]; found {
+		action.Resource.FeePaid = 0
```
Comment (oryband) [inline services/horizon/internal/actions_transaction.go]: nice use of `if ; ok` practice and also semantic name for `found` :+1:

File: services/horizon/internal/actions_transaction.go [hunk 5]
```diff
@@ -148,11 +151,27 @@ func (action *TransactionShowAction) JSON() error {
...
 }
 
+func validateFee(action *TransactionShowAction) {
+	mAccountAction := AccountShowAction{}
+	mWhiteListAccount := WhitelistAccount{}
+
+	_ = action.CoreQ().AllDataByAddress(&mAccountAction.CoreData, action.App.WhiteListAccount)
+	for _, d := range mAccountAction.CoreData {
+		mWhiteListAccount.WhitelistedAccounts[d.Key] = d.Value
+	}
```
Comment (oryband) [inline services/horizon/internal/actions_transaction.go]: - this causes every `transaction/...` request to make another DB query, fetching the large amount of data in the whitelist account, and then populate it in an object locally. we sure this is ok performance wise?
  - just a thought, but maybe we can populate once on demand (like on the first request depending on it) and then cache the whitelist for each of the last 10 or so blocks somewhere e.g. in the db.
  - maybe we should just implement this in core at a later time since whatever we do here is a hack and will never be the source of truth (where the truth is the data fetched from core)
- also i would add a comment explaining the logic being done here in this block since it's not trivial: from whta i understand it overrides the fee for the source account in this tx if it is whitelitsed, right?
- i'm not sure this actually solves the problem - the tx source account isn't always the one that is whitelisted. a transaction is exempt from fees if one of it's signatures is whitelisted. the source account signature must be included, but sometimes it's not the whitelisted signature. i think (but not sure) that in order to get this right you need to see if at least one of the signatures for this tx is in the whitelist account.

File: protocols/stellarcore/info_response.go [hunk 6]
```diff
@@ -4,12 +4,12 @@ package stellarcore
...
-		ProtocolVersion int        `json:"protocol_version"`
-		State           string     `json:"state"`
-		Ledger          LedgerInfo `json:"ledger"`
-
+		Build            string     `json:"build"`
+		Network          string     `json:"network"`
+		ProtocolVersion  int        `json:"protocol_version"`
+		State            string     `json:"state"`
+		Ledger           LedgerInfo `json:"ledger"`
+		WhiteListAccount string     `json:"whitelist"`
```
Comment (oryband) [inline protocols/stellarcore/info_response.go:12]: general comment about this PR not related to this line:
- there are four commits in this PR which have the exact same message. either squash them to a single one or change the messsages so it clearly explains what's being changed/added/removed in each commit
- also add a summary title + description to PR so we won't have to dig into every commit message to understand the general context and purpose of this PR
- please add unit tests

File: services/horizon/internal/actions_transaction.go [hunk 7]
```diff
@@ -211,7 +230,7 @@ func (action *TransactionCreateAction) loadResource() {
 	switch err := action.Result.Err.(type) {
 	case *txsub.FailedTransactionError:
 		rcr := horizon.TransactionResultCodes{}
-		resourceadapter.PopulateTransactionResultCodes(action.R.Context(), &rcr, err)
+		_ = resourceadapter.PopulateTransactionResultCodes(action.R.Context(), &rcr, err)
```
Comment (oryband) [inline services/horizon/internal/actions_transaction.go:213]: why is it needed to assign to an used var here?

File: services/horizon/internal/app.go [hunk 8]
```diff
@@ -25,9 +24,10 @@ import (
 	"github.com/kinecosystem/go/support/app"
 	"github.com/kinecosystem/go/support/db"
 	"github.com/kinecosystem/go/support/log"
+	"github.com/rcrowley/go-metrics"
```
Comment (oryband) [inline services/horizon/internal/app.go:27]: - this renames `metrics` to `go-metrics` without renaming where this package is called in this file. how did this pass compilation?
- if possible i think it's better to avoid making unnecessary changes to important files like `app.go`. we'll have to solve merge conflicts when fetching updates from stellar

File: services/horizon/internal/app.go [hunk 9]
```diff
@@ -25,9 +24,10 @@ import (
...
 	"github.com/throttled/throttled"
 	"golang.org/x/net/http2"
-	graceful "gopkg.in/tylerb/graceful.v1"
+	"gopkg.in/tylerb/graceful.v1"
```
Comment (oryband) [inline services/horizon/internal/app.go:30]: same as previous comment. also applies for all other files

File: services/horizon/internal/app.go [hunk 10]
```diff
@@ -48,6 +48,7 @@ type App struct {
 	ingester                     *ingest.System
 	reaper                       *reap.System
 	ticks                        *time.Ticker
+	WhiteListAccount             string
```
Comment (oryband) [inline services/horizon/internal/app.go]: - i don't think this member needs to be public. make it private
- add a smal comment explaining what this is and why is it here (for caching the whitelist address), regardless that the other members here are undocumented.

File: services/horizon/internal/actions_transaction.go (hunk 4 above)
Comment (oryband) [reply services/horizon/internal/actions_transaction.go]: this is opinionated but try to put some lines spaces between clauses according to context i.e. there's a function call, for loop, and if clause all squashed together one after the other. it's hard to read (for me)

**Review by oryband — CHANGES_REQUESTED** [review]

File: services/horizon/internal/actions_transaction.go (hunk 1 above)
Comment (Yogev911) [reply services/horizon/internal/actions_transaction.go:14]: gofmt.
let's stay aligned with golang standarts

File: services/horizon/internal/actions_transaction.go (hunk 3 above)
Comment (Yogev911) [reply services/horizon/internal/actions_transaction.go]: removed, created a new function with doc.

File: services/horizon/internal/actions_transaction.go (hunk 5 above)
Comment (Yogev911) [reply services/horizon/internal/actions_transaction.go]: solved, created updateWhiteListData function that called every new block (every 5 seconds) and caches the Data

File: services/horizon/internal/actions_transaction.go (hunk 7 above)
Comment (Yogev911) [reply services/horizon/internal/actions_transaction.go:213]: its error handling, the IDE screams when he sees an unhandled error, ( i have nothing to do with that error so I assign it to _

File: services/horizon/internal/app.go (hunk 8 above)
Comment (Yogev911) [reply services/horizon/internal/app.go:27]: IDE is doing it automatically, I guess that if the right use of it (gofmt)

File: services/horizon/internal/app.go (hunk 10 above)
Comment (Yogev911) [reply services/horizon/internal/app.go]: moved to Config

File: services/horizon/internal/db2/core/transaction.go [hunk 11]
```diff
@@ -35,10 +42,36 @@ func (tx *Transaction) EnvelopeXDR() string {
...
+	if tx.isTxWhitelisted(whiteListData) {
+		return int32(0)
+	}
 	return int32(tx.Envelope.Tx.Fee)
 }
 
+func (tx *Transaction) isTxWhitelisted(whiteListData map[string]string) bool {
+	// First check if the source account is a Whitelist account.
+	if _, found := whiteListData[tx.Envelope.Tx.SourceAccount.Address()]; found {
+		return true
```
Comment (avi-kik) [inline services/horizon/internal/db2/core/transaction.go]: This is not sufficient to prove it was whitelisted, even though it will be correct more often than not.

There are signature types other than ED25119 that may be used, which means that the source account may not have actually signed the tx.  This might be considered an oversight by the whitelist code in core (i.e. if the source is whitelisted, we shouldn't care about the signature), but at present this is the case.

Comment (Yogev911) [reply services/horizon/internal/db2/core/transaction.go]: Solved, added TODO and Jira ticket

File: services/horizon/internal/db2/core/transaction.go [hunk 12]
```diff
@@ -51,9 +51,10 @@ func (tx *Transaction) Fee(whiteListData map[string]string) int32 {
 
 func (tx *Transaction) isTxWhitelisted(whiteListData map[string]string) bool {
 	// First check if the source account is a Whitelist account.
-	if _, found := whiteListData[tx.Envelope.Tx.SourceAccount.Address()]; found {
-		return true
-	}
+	// TODO This check will be valid when this Jira task will be solved : https://kin.atlassian.net/browse/BC-814
```
Comment (oryband) [inline services/horizon/internal/db2/core/transaction.go]: the link here is private and nobody outside of kin ecosystem can read it. i think it's better to explain the issue here in length

**Review by oryband — DISMISSED** [review]
- tests are failing
- there are a lot of reudndant commits:
image
- there are commits from other branches (already in master) in this pr
image
- some commit messages are vague, please explain more precisely what changed in the commit
image

i'll review it once all of this is resolved

File: services/horizon/internal/ingest/main_test.go [hunk 13]
```diff
@@ -55,12 +55,12 @@ func TestTick(t *testing.T) {
 	sys := sys(tt, false)
 
 	// ingest by tick
-	s := sys.Tick()
+	s := sys.Tick("GBD53ST2UUC5VESGEFIPSNZPSVIHIVV5G3SFCCMGNWD7ETMDV3HNA2JV")
```
Comment (fuzzyami) [inline services/horizon/internal/ingest/main_test.go:58]: do we want to support a scenario where there is no white list?

Comment (Yogev911) [reply services/horizon/internal/ingest/main_test.go:58]: This text function expect to receive a string, and the design of our web uses whitelist.

Comment (fuzzyami) [reply services/horizon/internal/ingest/main_test.go:58]: let me rephrase my question: what happens if we dont configure a whitelist? does horizon crash?

**Review by fuzzyami — APPROVED** [review]
